mantidproject/mantid
# Add option for PolarizationCorrectionWildes to include output spin state in log

### Description of work

#### Summary of work

Adds a new boolean parameter to `PolarizationCorrectionWildes` that, when set to `True`, adds an entry to the sample log of each output child workspace specifying the spin state of its data. This new parameter is also exposed through the wrapper algorithm `PolarizationEfficiencyCor`.



Fixes #38264. 




#### Further detail of work


- The new parameter is set to `False` by default to maintain backwards compatibility.
- The intention is to add the same parameter to `PolarizationCorrectionFredrikze` very soon, so I haven't done anything programmatically in `PolarizationEfficiencyCor` to prevent the new `AddSpinStateToLog` parameter from being used when running a Fredrikze correction. For the time-being, the docs provide an explanation that the new parameter only works with the Wildes algorithm, which will obviously need to be updated once support has been extended to Fredrikze.
- This PR switches a number of string constants from using type `auto constexpr` to `static const std::string`. Having looked into it further, it seems that because `constexpr` is not implicitly convertible to a string it means that trying to pass it by reference creates a copy. As a result, it seemed to me that it would be better to use type `static const std::string` for string constants.
- I've added new unit tests for both `PolarizationCorrectionWildes` and `PolarizationEfficiencyCor`. The unit test classes for both of these algorithms contain quite a lot of code duplication and would benefit from some maintenance. I would prefer not to do that as part of this PR (due to the time required plus wanting to keep the PR as small as possible), so will open a couple of maintenance issues to address this.

### To test:



1) Run the following script. It will output a group of workspaces called "corrected". Check that the sample log of each child workspace **does not** contain an entry with the name `spin_state_ORSO`.
```
CreateSampleWorkspace(OutputWorkspace='out_00', Function='User Defined', UserDefinedFunction='name=Lorentzian, Amplitude=48000, PeakCentre=2.65, FWHM=1.2', XUnit='wavelength', NumBanks=1, BankPixelWidth=1, XMin=0, XMax=16.5, BinWidth=0.1)
CreateSampleWorkspace(OutputWorkspace='out_11', Function='User Defined', UserDefinedFunction='name=Lorentzian, Amplitude=47000, PeakCentre=2.65, FWHM=1.2', XUnit='wavelength', NumBanks=1, BankPixelWidth=1, XMin=0, XMax=16.5, BinWidth=0.1)
CreateSampleWorkspace(OutputWorkspace='out_10', Function='User Defined', UserDefinedFunction='name=Lorentzian, Amplitude=22685, PeakCentre=2.55, FWHM=0.6', XUnit='wavelength', NumBanks=1, BankPixelWidth=1, XMin=0, XMax=16.5, BinWidth=0.1)
CreateSampleWorkspace(OutputWorkspace='out_01', Function='User Defined', UserDefinedFunction='name=Lorentzian, Amplitude=22685, PeakCentre=2.55, FWHM=0.6', XUnit='wavelength', NumBanks=1, BankPixelWidth=1, XMin=0, XMax=16.5, BinWidth=0.1)
group = GroupWorkspaces(['out_00','out_11','out_10','out_01'])
group = ConvertUnits(group, "Wavelength")

ws1 = CreateWorkspace([0.01, 17.0, 34.0], [50000, 50000])
eff = JoinISISPolarizationEfficiencies(P1=ws1,   P2=ws1, F1=ws1, F2=ws1)
eff = ConvertUnits(eff, "Wavelength")
eff = RebinToWorkspace(eff, group[0], True)

corrected = PolarizationCorrectionWildes([group[0], group[1], group[2], group[3]], Efficiencies=eff)
```

2) Without clearing any of your workspaces, run the following:

```
with_log = PolarizationCorrectionWildes([group[0], group[1], group[2], group[3]], Efficiencies=eff, AddSpinStateToLog=True)
```
Check that the sample log of each child workspace in the "with_log" group **does** contain an entry with the name `spin_state_ORSO`. The log values should be as follows:

- `_++` workspace has value `pp`
- `_+-` workspace has value `pm`
- `_-+` workspace has value `mp`
- `_--` workspace has value `mm`





---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
